src/asl.ts

```diff
@@ -2774,37 +2772,64 @@ export class ASL {
         const ref = expr.lookup();
         if (ref === undefined) {
           throw new Error(`unresolved identifier: ${expr.name}`);
-        } else if (isParameterDecl(ref)) {
-          if (ref.parent !== predicate) {
-            throw new Error(
-              "cannot reference a ParameterDecl other than those in .filter((item, index) =>) in a JSONPath filter expression"
-            );
-          }
-          if (ref === ref.parent.parameters[0]) {
-            return "@";
-          } else if (ref === ref.parent.parameters[1]) {
-            throw new Error(
-              "the 'index' parameter in a .filter expression is not supported"
-            );
-          } else {
+        }
+        return resolveRef(ref);
+        function resolveRef(ref: Decl): string {
+          if (isParameterDecl(ref)) {
+            if (ref.parent !== predicate) {
+              throw new Error(
+                "cannot reference a ParameterDecl other than those in .filter((item, index) =>) in a JSONPath filter expression"
+              );
+            }
+            if (ref === ref.parent.parameters[0]) {
+              return "@";
+            } else if (ref === ref.parent.parameters[1]) {
+              throw new Error(
+                "the 'index' parameter in a .filter expression is not supported"
+              );
+            } else {
+              throw new Error(
+                "the 'array' parameter in a .filter expression is not supported"
+              );
+            }
+          } else if (isBindingElem(ref)) {
+            if (isArrayBinding(ref.parent)) {
+              return `${resolveRef(
+                ref.parent.parent
+              )}[${ref.parent.bindings.indexOf(ref)}]`;
+            }
+
+            const propName = ref.propertyName
+              ? isIdentifier(ref.propertyName)
+                ? ref.propertyName.name
+                : isStringLiteralExpr(ref.propertyName)
+                ? ref.propertyName.value
+                : evalToConstant(ref.propertyName)?.constant
+              : isIdentifier(ref.name)
+              ? ref.name.name
+              : undefined;
+
+            if (!propName) {
+              throw new SynthError(
+                ErrorCodes.StepFunctions_property_names_must_be_constant
+              );
+            }
+
+            return `${resolveRef(ref.parent.parent)}[${propName}]`;
+          } else if (
+            isVariableDecl(ref) ||
+            isFunctionDecl(ref) ||
+            isClassDecl(ref) ||
+            isClassMember(ref) ||
+            isGetAccessorDecl(ref) ||
+            isSetAccessorDecl(ref)
+          ) {
             throw new Error(
-              "the 'array' parameter in a .filter expression is not supported"
+              `cannot reference a ${ref.kind} within a JSONPath .filter expression`
```

Comment:
will have to change this to `ref.kindName` when you merge.

Reply:
:+1: you may have changed it, we'll see how the merge goes (this was existing code moved a bit)